src/monodroid/jni/monodroid-glue.c

```diff
@@ -1278,92 +1278,271 @@ gc_is_bridge_object (MonoObject *object)
+{
+	MonoJavaGCBridgeInfo *bridge_info = NULL, *reffed_bridge_info = NULL;
+	jobject handle, reffed_handle;
+
+	if (!load_reference_target (target, &bridge_info, &handle))
+		return FALSE;
+
+	if (!load_reference_target (reffed_target, &reffed_bridge_info, &reffed_handle))
+		return FALSE;
+
+	mono_bool success = add_reference_jobject (env, handle, reffed_handle);
+
+	// Flag MonoObjects so they can be cleared in gc_cleanup_after_java_collection.
+	// Java temporaries do not need this because the entire GCUserPeer is discarded.
+	if (success && target.is_mono_object) {
+		int ref_val = 1;
+		mono.mono_field_set_value (target.obj, bridge_info->refs_added, &ref_val);
+	}
+
+#if DEBUG
+	if (gc_spew_enabled) {
+		char *description = describe_target (target),
+			 *reffed_description = describe_target (reffed_target);
+
+		if (success)
+			log_warn (LOG_GC, "Added reference for %s to %s", description, reffed_description);
+		else
+			log_error (LOG_GC, "Missing monodroidAddReference method for %s", description);
+
+		free (description);
+		free (reffed_description);
+	}
+#endif
+
+	return success;
+}
+
+// Create a target
+static AddReferenceTarget
+target_from_mono_object (MonoObject *obj)
+{
+	AddReferenceTarget result;
+	result.is_mono_object = TRUE;
+	result.obj = obj;
+	return result;
+}
+
+// Create a target
+static AddReferenceTarget
+target_from_jobject (jobject jobj)
+{
+	AddReferenceTarget result;
+	result.is_mono_object = FALSE;
+	result.jobj = jobj;
+	return result;
+}
+
+/* During the xref phase of gc_prepare_for_java_collection, we need to be able to map bridgeless
+ * SCCs to their index in temporary_peers. Because for all bridgeless SCCs the num_objs field of
+ * MonoGCBridgeSCC is known 0, we can temporarily stash this index as a negative value in the SCC
+ * object. This does mean we have to erase our vandalism at the end of the function.
+ */
+static int
+scc_get_stashed_index (MonoGCBridgeSCC *scc)
+{
+	assert ( (scc->num_objs < 0) || !"Attempted to load stashed index from an object which does not contain one." );
+	return -scc->num_objs - 1;
+}
+
+static void
+scc_set_stashed_index (MonoGCBridgeSCC *scc, int index)
+{
+	scc->num_objs = -index - 1;
+}
+
+// Extract the root target for an SCC. If the SCC has bridged objects, this is the first object. If not, it's stored in temporary_peers.
+static AddReferenceTarget
+target_from_scc (MonoGCBridgeSCC **sccs, int idx, JNIEnv *env, jobject temporary_peers)
+{
+	MonoGCBridgeSCC *scc = sccs [idx];
+	if (scc->num_objs > 0)
+		return target_from_mono_object (scc->objs [0]);
+
+	jobject jobj = (*env)->CallObjectMethod (env, temporary_peers, ArrayList_get, scc_get_stashed_index (scc));
+	return target_from_jobject (jobj);
+}
+
+// Must call this on any AddReferenceTarget returned by target_from_scc once done with it
+static void
+target_release (JNIEnv *env, AddReferenceTarget target)
+{
+	if (!target.is_mono_object)
+		(*env)->DeleteLocalRef (env, target.jobj);
+}
```

Comment:
Wait, hold on: No, we should not set it to NULL, because target is not being passed by reference. "target" is here a function temporary.